File: Baekjoon/Floyd-Warshall/14938.py
import sys

input = sys.stdin.readline
INF = int(1e9)

n, m, r = map(int, input().split()) # 지역, 수색 범위, 길의 개수

items = [0] + list(map(int, input().split()))

graph = [[INF] * (n+1) for _ in range(n+1)]

for i in range(1, n+1):
	graph[i][i] = 0

for _ in range(r):
	a, b, l = map(int, input().split())
	graph[a][b] = l
	graph[b][a] = l

for k in range(1, n+1):
	for i in range(1, n+1):
		for j in range(1, n+1):
			graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j])

max_count = 0
for i in range(1, n+1):
	count = 0
	for j in range(1, n+1):
		if graph[i][j] <= m:
			count += items[j]
	max_count = max(max_count, count)
print(max_count)

File: Baekjoon/Floyd-Warshall/14938.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int INF = 1e9;

int main() {
	int n, m, r;
	cin >> n >> m >> r;

	vector<int> items(n+1, 0);
	for (int i = 1; i <= n; i++) {
		cin >> items[i];
	}

	// graph = [[INF] * (n+1) for _ in range(n+1)]
	vector<vector<int> > graph(n+1, vector<int>(n+1, INF));
	for (int i = 1; i <= n; i++) {
		graph[i][i] = 0;
	}
	for (int i = 0; i < r; i++) {
		int a, b, l;
		cin >> a >> b >> l;
		graph[a][b] = l;
		graph[b][a] = l;
	}

	for (int k = 1; k <= n; k++) {
		for (int i = 1; i <= n; i++) {
			for (int j = 1; j <= n; j++) {
				graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j]);
			}
		}
	}

	int max_count = 0;
	for (int i = 1; i <= n; i++) {
		int count = 0;
		for (int j = 1; j <= n; j++) {
			if (graph[i][j] <= m) {
				count += items[j];
			}
		}
		max_count = max(max_count, count);
	}
	cout << max_count << endl;

	return 0;
}
